src/Libraries/Microsoft.Extensions.AI.OpenAI/OpenAIResponseChatClient.cs

```diff
@@ -0,0 +1,582 @@
+                    {
+                        functionCallInfos[itemUpdate.ItemIndex] = callInfo = new();
+                    }
+
+                    callInfo.CallId = functionCallItem.CallId;
+                    callInfo.Name = functionCallItem.FunctionName;
+                    continue;
+                }
+            }
+
+            // Handle content updates.
+            if (streamingUpdate is StreamingResponseContentPartDeltaUpdate contentUpdate)
+            {
+                // Update our knowledge of function call requests.
+                if (contentUpdate.FunctionArguments is string argsUpdate)
+                {
+                    functionCallInfos ??= [];
+                    if (!functionCallInfos.TryGetValue(contentUpdate.ItemIndex, out FunctionCallInfo? callInfo))
+                    {
+                        functionCallInfos[contentUpdate.ItemIndex] = callInfo = new();
+                    }
+
+                    _ = (callInfo.Arguments ??= new()).Append(argsUpdate);
+                }
+
+                // If there's any text content, return it.
+                if (!string.IsNullOrEmpty(contentUpdate.Text))
+                {
+                    yield return new(role, contentUpdate.Text)
+                    {
+                        CreatedAt = createdAt,
+                        ModelId = modelId,
+                        RawRepresentation = streamingUpdate,
+                        ResponseId = responseId,
+                    };
+                }
+
+                continue;
+            }
+        }
+
+        // Now that we've received all updates and yielded all content,
+        // yield a final update with any remaining information.
+        ChatResponseUpdate update = new()
+        {
+            ResponseId = responseId,
+            CreatedAt = createdAt,
+            FinishReason = finishReason ?? (functionCallInfos is not null ? ChatFinishReason.ToolCalls : ChatFinishReason.Stop),
+            ModelId = modelId,
+            Role = role,
+        };
+
+        if (usage is not null)
+        {
+            update.Contents.Add(new UsageContent(usage));
+        }
+
+        if (functionCallInfos is not null)
+        {
+            foreach (var entry in functionCallInfos)
+            {
+                FunctionCallInfo fci = entry.Value;
+                if (!string.IsNullOrWhiteSpace(fci.Name))
+                {
+                    update.Contents.Add(
+                        FunctionCallContent.CreateFromParsedArguments(
+                            fci.Arguments?.ToString() ?? string.Empty,
+                            fci.CallId ?? string.Empty,
+                            fci.Name!,
+                            static json => JsonSerializer.Deserialize(json, OpenAIJsonContext.Default.IDictionaryStringObject)!));
+                }
+            }
+        }
+
+        yield return update;
```

Comment:
I expect this may be unavoidable, but it looks like it could change the order of function calls relative to text content within the response stream. All the text content will be yielded first, followed by all the function calls, even if they were interleaved in some way in the source data.

Is this because we can't know when a given function call is "complete" and could be yielded? Is there any possibility that a function call can be spread out and be interleaved with text contents or other function calls, or do we know that once we receive the first update for a given function call, all the consecutive ones will be for the same call until that call's data is fully consumed?

Reply:
> Is this because we can't know when a given function call is "complete" and could be yielded?

Yes. @trrwilson, is there any way through the API to determine when we get the last part of a function call? There wasn't with chat completion, and I didn't see anything obvious for responses, either.